### PR TITLE
Move LowerGpuRt into LGC

### PR DESCRIPTION
**Pull request by LLJJDD**
With some tweak to adapt the change.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit d15e7f25148284916e5869797ff4daf93173ba9c
<details><summary>CTS tests (Failed: 1645/138443)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 33639/69228 (48.6%)</li>
<li><details><summary>Failed: 1587/69228 (2.3%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.buffer.suballocation.transfer_src.transfer_dst.uniform_texel.storage_texel.uniform.storage.create.binding_residency
Stack trace: Flake

FAILURE: dEQP-VK.api.buffer.suballocation.uniform_texel.uniform.index.create.binding
Stack trace: Flake

FAILURE: dEQP-VK.api.buffer.suballocation.uniform_texel.uniform.index.indirect.create.binding_aliased
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.blit_image.all_formats.generate_mipmaps.from_previous_level.layercount_6.r8g8_snorm.general_optimal_nearest
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.buffer_to_depthstencil.buffer_offset_d16_unorm_s8_uint_D_S
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.core.blit_image.simple_tests.scaling_whole2.b8g8r8a8_unorm_linear
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.core.blit_image.simple_tests.without_scaling_partial_3d.b8g8r8a8_unorm_linear
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.core.buffer_to_buffer_with_offset.6_1
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.core.depth_stencil_msaa_copy.whole.d16_unorm_s8_uint_general_general_D_8_bit_bind_offset
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.blit_image.simple_tests.without_scaling_partial_3d.linear
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.image_to_image.dimensions.src4x8192_dst4x8192.r4g4_unorm_pack8.r8_srgb.optimal_optimal
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.image_to_image.dimensions.src6x16384_dst6x16384.r32_sfloat.r32_sfloat.optimal_optimal
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image_one_region.4_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.external.fence.sync_fd.export_multiple_times_temporary
Stack trace: Fail

FAILURE: dEQP-VK.api.image_clearing.core.clear_color_attachment.single_layer.r8_srgb_64x11
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_color_image.1d.linear.remaining_array_layers_twostep.r16g16b16a16_sfloat_200x1
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_color_image.1d.optimal.remaining_array_layers_twostep.r8g8b8a8_sint_200x1
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_color_image.2d.optimal.remaining_array_layers.r8g8_sint_64x11
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.partial_clear_color_attachment.cube_layers.r16g16_sint
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.partial_clear_color_attachment.single_layer.a4b4g4r4_unorm_pack16_64x11_clamp_input
Stack trace: Flake
...
```
</p></details></li>

<li>Not Supported: 34002/69228 (49.1%)</li><li>Warnings: 0/69228 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35184/69215 (50.8%)</li>
<li><details><summary>Failed: 58/69215 (0.1%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.ray_query.builtin.getintersectionfrontfaceCandidate.geom.triangles
Stack trace: Fail

FAILURE: dEQP-VK.ray_query.builtin.getintersectionfrontfaceCommitted.frag.triangles
Stack trace: Fail

FAILURE: dEQP-VK.ray_query.builtin.getintersectiongeometryindexCommitted.frag.aabbs
Stack trace: Fail

FAILURE: dEQP-VK.ray_query.builtin.getintersectiongeometryindexCommitted.geom.aabbs
Stack trace: Fail

FAILURE: dEQP-VK.ray_query.builtin.getworldrayorigin.geom.triangles
Stack trace: Fail

FAILURE: dEQP-VK.ray_query.builtin.getworldrayorigin.vert.triangles
Stack trace: Fail

FAILURE: dEQP-VK.ray_query.builtin.rayqueryterminate.geom.aabbs
Stack trace: Fail

FAILURE: dEQP-VK.ray_query.helper_invocations.gpu.coarse.cubic_quadratic.32x64.64x32
Stack trace: Fail

FAILURE: dEQP-VK.ray_query.helper_invocations.gpu.coarse.cubic_quadratic.32x64.64x64
Stack trace: Fail

FAILURE: dEQP-VK.ray_query.helper_invocations.gpu.coarse.cubic_quadratic.64x64.64x32
Stack trace: Fail

FAILURE: dEQP-VK.ray_query.helper_invocations.gpu.coarse.cubic_quadratic.64x64.64x64
Stack trace: Fail

FAILURE: dEQP-VK.ray_query.helper_invocations.gpu.coarse.linear_cubic.32x64.64x32
Stack trace: Fail

FAILURE: dEQP-VK.ray_query.helper_invocations.gpu.coarse.linear_cubic.32x64.64x64
Stack trace: Fail

FAILURE: dEQP-VK.ray_query.helper_invocations.gpu.coarse.linear_cubic.64x64.64x32
Stack trace: Fail

FAILURE: dEQP-VK.ray_query.helper_invocations.gpu.coarse.linear_cubic.64x64.64x64
Stack trace: Fail

FAILURE: dEQP-VK.ray_query.helper_invocations.gpu.coarse.linear_quadratic.32x64.64x32
Stack trace: Fail

FAILURE: dEQP-VK.ray_query.helper_invocations.gpu.coarse.linear_quadratic.32x64.64x64
Stack trace: Fail

FAILURE: dEQP-VK.ray_query.helper_invocations.gpu.coarse.linear_quadratic.64x64.64x32
Stack trace: Fail

FAILURE: dEQP-VK.ray_query.helper_invocations.gpu.coarse.linear_quadratic.64x64.64x64
Stack trace: Fail

FAILURE: dEQP-VK.ray_query.helper_invocations.gpu.fine.cubic_quadratic.32x64.64x32
Stack trace: Fail
...
```
</p></details></li>

<li>Not Supported: 33973/69215 (49.1%)</li><li>Warnings: 0/69215 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by LLJJDD** [comment]
retest this please

File: lgc/patch/LowerGpuRt.cpp [hunk 1]
```diff
@@ -117,26 +118,48 @@ unsigned LowerGpuRt::getWorkgroupSize() const {
...
-  auto ldsStack = new GlobalVariable(*m_module, m_stackTy, false, GlobalValue::ExternalLinkage, nullptr,
-                                     RtName::LdsStack, nullptr, GlobalValue::NotThreadLocal, 3);
-
-  ldsStack->setAlignment(MaybeAlign(4));
-  m_stack = ldsStack;
+// @param [in/out] module : LLVM module to be run on
+void LowerGpuRt::createGlobalStack(Module &module) {
+
+  struct Payload {
+    bool needGlobalStack;
```
Comment (tsymalla-AMD) [inline lgc/patch/LowerGpuRt.cpp:132]: If it is not against common practice, I'd prefer to initialize the members inside the struct

File: lgc/patch/LowerGpuRt.cpp [hunk 2]
```diff
@@ -117,26 +118,48 @@ unsigned LowerGpuRt::getWorkgroupSize() const {
...
+    bool needExtraStack;
+  };
+  Payload payload = {false, false};
+  static auto visitor = llvm_dialects::VisitorBuilder<Payload>()
+                            .setStrategy(llvm_dialects::VisitorStrategy::ByFunctionDeclaration)
+                            .add<GpurtStackWriteOp>([](auto &payload, auto &op) {
```
Comment (tsymalla-AMD) [inline lgc/patch/LowerGpuRt.cpp:138]: Another way to do this would be to use `addSet`, but it would require an additional cast in the callback.

**Comment by tsymalla-AMD** [comment]
Some comments, apart from that it looks good. Are the Jenkins / CTS failures related?

File: lgc/patch/LowerGpuRt.cpp (hunk 1 above)
Comment (LLJJDD) [reply lgc/patch/LowerGpuRt.cpp:130]: I do see some other places use this form, e.g.
https://github.com/GPUOpen-Drivers/llpc/blob/8e2673136e10557738b047dd1b8584dde1ba9590/lgc/patch/PatchEntryPointMutate.cpp#L271

File: lgc/patch/LowerGpuRt.cpp (hunk 2 above)
Comment (LLJJDD) [reply lgc/patch/LowerGpuRt.cpp:136]: Thanks! It seems it makes minor difference so I keep as-is ;)

**Comment by LLJJDD** [comment]
> Some comments, apart from that it looks good. Are the Jenkins / CTS failures related?

Yes, hopefully my latest commit fixes them.

File: llpc/context/llpcPipelineContext.cpp [hunk 3]
```diff
@@ -342,6 +342,10 @@ Options PipelineContext::computePipelineOptions() const {
   options.enablePrimGeneratedQuery = getPipelineOptions()->enablePrimGeneratedQuery;
   options.enableFragColor = getPipelineOptions()->enableFragColor;
 
+  options.rtBoxSortHeuristicMode = m_rtState.boxSortHeuristicMode;
+  options.rtStaticPipelineFlags = m_rtState.staticPipelineFlags;
+  options.rtTriCompressMode = m_rtState.triCompressMode;
```
Comment (tsymalla-AMD) [inline llpc/context/llpcPipelineContext.cpp:347]: Do you need to default-initialize useGpurt here as well?

Comment (nhaehnle) [reply llpc/context/llpcPipelineContext.cpp:347]: The `= {}` initialization at the top should take care of that, right?

**Review by nhaehnle — APPROVED** [review]
Following yesterday's discussion, luckily @mbrkusanin is going to work on moving the descriptor load lowering out of the builder replayer. Once that's done, it should be possible to remove this workaround and move LowerGpuRt to after the BuilderReplayer and in fact after LowerRaytracingPipeline, where it should be in the long run.

Maybe you can double-check Thomas' question, but apart from that this LGTM.

File: llpc/context/llpcPipelineContext.cpp (hunk 3 above)
Comment (tsymalla-AMD) [reply llpc/context/llpcPipelineContext.cpp:347]: Yes. My question is based on the fact that we initialize other members with a constant value here as well, so it is merely a question about coding style.

Comment (LLJJDD) [reply llpc/context/llpcPipelineContext.cpp:347]: I think we don't need to because it is always false (unknown) at here.

**Pull request merged**